### DISCUSSION TITLE
Custom run launchers

### DISCUSSION THREAD

**Discussion by johannkm** [dagster-io/dagster#15856] [Q&A] [answered]
How can I write and deploy a custom run launcher?

**Comment by johannkm** [comment, marked as answer]
**Note:**  this guide applies to OSS. For Dagster cloud hybrid, it's possible to customize the run launcher used in the agent but requires setting a custom [UserCodeLauncher](https://github.com/dagster-io/dagster-cloud/blob/main/dagster-cloud/dagster_cloud/workspace/user_code_launcher/user_code_launcher.py). Reach out for help on this.

---

Per https://docs.dagster.io/deployment/run-launcher#custom-run-launchers, custom run launchers make sense when
- you are launching on to a platform the Dagster supports
- you have custom logic for allocating resources/where to launch runs/etc.

To use one:
1. Write a python class that implements the abstract [RunLauncher class](https://github.com/dagster-io/dagster/blob/master/python_modules/dagster/dagster/_core/launcher/base.py#L0-L1). It's a good idea to look at examples like the DockerRunLauncher and K8sRunLauncher.
2. Create a Python module that exports this run launcher class.
3. Install the custom module with your dagster deployment. The custom run launcher needs to be present in the containers for the dagster webserver and the dagster daemon.  If you're currently using the provided Docker images for these, you'll need to start building your own Docker images instead. You can use the provided image as a base, then install your library. Most often, this will require copying the module files in to the image, then running `pip install -e` on it.
5. Configure your instance to use the launcher. Run launchers are configured in `dagster.yaml`:
    ```yaml
      run_launcher:
        module: <your module>
        class: <your run launcher class>
        config:
          <any config if you set that up>
    ```
    If you're using the OSS Helm chart, you'll set in `values.yaml`: `runLauncher.type: CustomRunLauncher` and the module/class at `runLauncher.customRunLauncher.<your stuff>`

**Reply by cyberjar09** [reply]
thanks for the guide, I have followed this and created my custom run launcher. 🙇 

I wanted to ask about the `terminate()` invocation 

here is my sample code (just logging for now to ensure I can see the triggers happen)

```python
from dagster._core.launcher import LaunchRunContext, RunLauncher


class MyCustomRunLauncher(RunLauncher):
    def __init__(self):
        super().__init__()

    @classmethod
    def config_type(cls):
        return {}

    @classmethod
    def from_config_value(cls, inst_data, config_value):
        return cls()

    def launch_run(self, context: LaunchRunContext) -> None:
        print("+++++ context.dagster_run: ", context.dagster_run)

    def terminate(self, run_id):
        print(">>>>> terminate run_id: " + run_id)

``` 

as mentioned above, I was testing this and the `launch_run` trigger is working as expected, but I cant seem to trigger the `terminate()` invocation. I tried a few ways of terminating a run from the webserver:
- manually terminate from the run page
- force terminate
- terminate from the list of runs by ticking the checkbox

can you pls point me in the right direction? what could I be doing wrong? 

thanks 😄

**Reply by johannkm** [reply]
What run status is the run in when you're trying to cancel? 

Here's the graphql mutation implementation: https://github.com/dagster-io/dagster/blob/master/python_modules/dagster-graphql/dagster_graphql/implementation/execution/__init__.py#L81

The `run_coordinator.cancel_run` call invokes `run_launcher.terminate`. If the run isn't in STARTED state, then we won't call down to the run launcher.

**Reply by cyberjar09** [reply]
- thanks for the guidance @johannkm , I was able to see the issue when I triggered the gql termination. 
  - for some reason the stack trace was being swallowed when I triggered it from the webserver via clicking
- I was able to fix the issue on my end, and am able to trigger a termination via webserver now
  - but I observed something odd: upon termination, my custom run launcher does the needful and the run stops, but the status in the webserver remains "Canceling" and it does not reflect the terminated state. 🤔
  - ~~is there supposed to be some return type from my custom run launcher to inform the calling code in webserver?~~
  - edit: after few attempts of testing, it seems dagster termination is flakey. the cancelation is behaving as expected but inconsistently. Some runs it updates normally, others it wont update and eventually there is a timeout which forces the state to move to canceled.

**Comment by johannkm** [comment]
One case for a custom run launcher is if you want to use 2 or more of the provided run launchers, and to switch between them based on job name, a tag, etc.

A rough sketch of how to do this:
In your `__init__`, instantiate whichever other launchers you'll use. Add a `register_instance` method (see the class `MayHaveInstanceWeakref`, which the RunLauncher base class inherits) that calls the same method on each of the underlying launchers. Then for each of the methods like `launch_run`, pass the call on to whichever launcher you want to use for the given run.

**Reply by andrescolodrero** [reply]
Hi @johannkm , im having a very strange behaviour runing a dagster asset on k8srunlauncher [19429](https://github.com/dagster-io/dagster/issues/19429)
I dont understand quite well why this is happening, maybe the way k8srunlauncher is implemented.
for testing, can i have a custom launcher that run the code in the code location server? or multiprocess?

**Reply by noam-jacobson** [reply]
@johannkm Thanks for the initial sketch. I got a basic working of just running the DefaultRunLauncher. I want to implement the other Launchers.
How would I go about passing the tags from jobs and configurations from the yml files?

**Comment by ThomAub** [comment]
Hello,

Thank you for the extra documentation and advice. 
We've just released the first open-source version of our custom `RunLauncher` for Nomad, [dagster-nomad](https://github.com/PayLead/dagster-nomad). This was done by examining the `DockerRunLauncher`, `K8sRunLauncher`, and `EcsRunLauncher`.

I'm still trying to determine the best way to handle the ability to access different resources and job based on various code locations. 
I welcome any ideas, improvements, and suggestions for new unit/integration tests. 

Also, do you know if there's a template for Dagster third-party libraries to standardize definitions and the release process?

**Reply by johannkm** [reply]
This looks great! We don't have guidance so far, but I think we should at least start a list of community integrations etc.
I've forwarded this to our dev rel team
